python/src/open_discourse/01_data/03_download_raw_data_electoral_term_20.py

```diff
@@ -18,51 +18,81 @@
+            print(msg)
 
-    print("Done.")
 
-    for link in tqdm(
-        xml_links,
-        desc=f"Download XML-files for term {election_period['election_period']}...",
-    ):
-        url = "https://www.bundestag.de" + link.get("href")
-        page = requests.get(url, headers={"User-Agent": "Mozilla/5.0"})
-        session = regex.search(r"\d{5}(?=\.xml)", url).group(0)
+if __name__ == "__main__":
+    logger = MockLogger()
+    logger.info(
+        f"Begin task 01_03, download raw data for electoral term(s) {list(election_periods.keys())} ..."
+    )
+    for election_period in election_periods:
+        logger.info(
+            f"Script 01_03. Scraping links for term {election_period['election_period']}..."
+        )
 
-        with open(OUTPUT_PATH / (session + ".xml"), "w") as file:
-            file.write(
-                regex.sub(
-                    "</sub>",
-                    "",
-                    regex.sub("<sub>", "", page.content.decode("utf-8")),
+        offset = 0
+        xml_links = []
+
+        while True:
+            try:
+                URL = election_period["url"].format(str(offset))
+                page = requests.get(URL, headers={"User-Agent": "Mozilla/5.0"})
+                soup = BeautifulSoup(page.text, "html.parser")
+                # scrape for links
+                current_links = list(
+                    soup.find_all("a", attrs={"href": regex.compile("xml$")})
+                )
+                if len(current_links) != 0:
+                    xml_links += current_links
+                    offset += len(current_links)
+                else:
+                    break
+            except Exception as e:
+                logger.error(
+                    f"Error getting links for parliamentary sessions from {URL}, {e}"
                 )
-            )
-        time.sleep(1)
+                break
+        print("Done.")
 
-assert OUTPUT_PATH.exists()
-assert (
-    len(list(OUTPUT_PATH.glob("*.xml"))) > 0
-), "At least one XML file should be downloaded."
+        for link in tqdm(
+            xml_links,
+            desc=f"Download XML-files for term {election_period['election_period']}...",
+        ):
+            try:
+                url = "https://www.bundestag.de" + link.get("href")
+                page = requests.get(url, headers={"User-Agent": "Mozilla/5.0"})
+                session = regex.search(r"\d{5}(?=\.xml)", url).group(0)
 
-print("Script 01_03 done.")
+                with open(OUTPUT_PATH / (session + ".xml"), "w") as file:
+                    file.write(
+                        regex.sub(
+                            "</sub>",
+                            "",
+                            regex.sub("<sub>", "", page.content.decode("utf-8")),
+                        )
+                    )
+                time.sleep(1)
+            except Exception as e:
+                logger.error(f"Error downloading xml file from {url}: {e}")
+    try:
```

Comment:
this does not really make sense. Why catch the assert error when you are still raising an error in the end? Just to log the error? The assert statement is already printing a description when it fails so I think that is safe enough and the pipeline should not continue if we are missing data, in my opinion. But that is a point we can discuss: At the moment my assert statements will end the pipeline even if only one file is missing. That might be a bad idea so we should refactor this into a proper test that logs whenever a file could not be processed. Anyway, that was a very simple test at the beginning. Once we have proper logging for all the steps and functions we don't need this assert statement any longer as every function will log any error that occurs during processing.

Reply:
Yes, that is the discussion I wanted to kick off: If we implement logging with a logger class, imo all relevant info should end up in the logs (and possibly stdout). So here I catch the error to log it, and then raise it for the purpose of telling the Makefile this task failed since we currently track the information on that through *.done files.

Reply:
Totally fine for the XML part, I was talking about the assert statement part ;) But I agree, should be logged properly!